doc/man1/restorepalette.1
.TH restorepalette 1 "2 Aug 1997" "Svgalib (>= 1.2.11)" "Svgalib User Manual"
.SH NAME
restorepalette \- set the color palette for textmode.
.SH SYNOPSIS

.BI "restorepalette [" filename ]

.SH DESCRIPTION
.B restorepalette
without arguments sets the standard VGA palette. This can be useful
if it is somehow messed up.

With a
.I filename
argument a custom palette can be loaded (feature added by Charles Blake
<chuckb@alice.wonderland.caltech.edu>).

This allow a user to set up a file that looks like this one
(These color map definitions are the same as the default VGA ones.
Alter to suite personal tastes). The first column contains the number of
the color to set, then follow three integers in range 0 - 63 (lowest to
highest intensity) for red, green, blue.

.RS
.B " 0  0  0  0   # black"
.br
.B " 1  0  0 42   # blue"
.br
.B " 2  0 42  0   # green"
.br
.B " 3  0 42 42   # cyan"
.br
.B " 4 42  0  0   # red"
.br
.B " 5 42  0 42   # magenta"
.br
.B " 6 42 21  0   # brown"
.br
.B " 7 42 42 42   # white"
.br
.B " 8 21 21 21   # bright black"
.br
.B " 9 21 21 63   # bright blue"
.br
.B "10 21 63 21   # bright green"
.br
.B "11 21 63 63   # bright cyan"
.br
.B "12 63 21 21   # bright red"
.br
.B "13 63 21 63   # bright magenta"
.br
.B "14 63 63 21   # bright brown"
.br
.B "15 63 63 63   # bright white"
.RE

The inline comments are the
only kind of allowed, as I use a little
.BR fscanf (3)
trick to get them. Blank lines
are ok, but not pure comment lines. See the comments in my code, also.

This allows people to set up custom palettes for use in virtual console text
modes. I use it all the time. When combined with a color-syntax editor like
jed-0.97+ or color-ls, etc., being able to choose your own text-mode palette is
quite a bonus. I set mine up via
.B restorepalette /etc/palette
in my
.IR /etc/rc .
If the program is given the correct permissions, then individual users can
have
.B restorepalette ~/.palette
or some such thing in their shell startup
files.
Of course, it shouldn't be done when starting remote shells or when
under X, so some kind of test that
.B TERM
is a virtual console is needed for that case.

.SH SEE ALSO

.BR svgalib (7),
.BR vgagl (7),
.BR libvga.config (5),
.BR vga_setpalvec (3),
.BR dumpreg (1),
.BR convfont (1),
.BR fix132x43 (1),
.BR restorefont (1),
.BR restoretextmode (1),
.BR runx (1),
.BR savetextmode (1),
.BR setmclk (1),
.BR textmode (1).
.SH AUTHOR

This manual page was edited by Michael Weller <eowmob@exp-math.uni-essen.de>. The
exact source of the referenced utility as well as of the original documentation is
unknown.

It is very likely that both are at least to some extent are due to
Harm Hanemaayer <H.Hanemaayer@inter.nl.net>.

Occasionally this might be wrong. I hereby
asked to be excused by the original author and will happily accept any additions or corrections
to this first version of the svgalib manual.


utils/restorepalette.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <vga.h>
#include <errno.h>
#include <string.h>            /* strerror */
#include "../src/libvga.h"


/* default palette values */

static char default_red[256]
=
{0, 0, 0, 0, 42, 42, 42, 42, 21, 21, 21, 21, 63, 63, 63, 63,
 0, 5, 8, 11, 14, 17, 20, 24, 28, 32, 36, 40, 45, 50, 56, 63,
 0, 16, 31, 47, 63, 63, 63, 63, 63, 63, 63, 63, 63, 47, 31, 16,
 0, 0, 0, 0, 0, 0, 0, 0, 31, 39, 47, 55, 63, 63, 63, 63,
 63, 63, 63, 63, 63, 55, 47, 39, 31, 31, 31, 31, 31, 31, 31, 31,
 45, 49, 54, 58, 63, 63, 63, 63, 63, 63, 63, 63, 63, 58, 54, 49,
 45, 45, 45, 45, 45, 45, 45, 45, 0, 7, 14, 21, 28, 28, 28, 28,
 28, 28, 28, 28, 28, 21, 14, 7, 0, 0, 0, 0, 0, 0, 0, 0,
 14, 17, 21, 24, 28, 28, 28, 28, 28, 28, 28, 28, 28, 24, 21, 17,
 14, 14, 14, 14, 14, 14, 14, 14, 20, 22, 24, 26, 28, 28, 28, 28,
 28, 28, 28, 28, 28, 26, 24, 22, 20, 20, 20, 20, 20, 20, 20, 20,
 0, 4, 8, 12, 16, 16, 16, 16, 16, 16, 16, 16, 16, 12, 8, 4,
 0, 0, 0, 0, 0, 0, 0, 0, 8, 10, 12, 14, 16, 16, 16, 16,
 16, 16, 16, 16, 16, 14, 12, 10, 8, 8, 8, 8, 8, 8, 8, 8,
 11, 12, 13, 15, 16, 16, 16, 16, 16, 16, 16, 16, 16, 15, 13, 12,
 11, 11, 11, 11, 11, 11, 11, 11, 0, 0, 0, 0, 0, 0, 0, 0};
static char default_green[256]
=
{0, 0, 42, 42, 0, 0, 21, 42, 21, 21, 63, 63, 21, 21, 63, 63,
 0, 5, 8, 11, 14, 17, 20, 24, 28, 32, 36, 40, 45, 50, 56, 63,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 31, 47, 63, 63, 63, 63,
 63, 63, 63, 63, 63, 47, 31, 16, 31, 31, 31, 31, 31, 31, 31, 31,
 31, 39, 47, 55, 63, 63, 63, 63, 63, 63, 63, 63, 63, 55, 47, 39,
 45, 45, 45, 45, 45, 45, 45, 45, 45, 49, 54, 58, 63, 63, 63, 63,
 63, 63, 63, 63, 63, 58, 54, 49, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 7, 14, 21, 29, 28, 28, 28, 28, 28, 28, 28, 28, 21, 14, 7,
 14, 14, 14, 14, 14, 14, 14, 14, 14, 17, 21, 24, 28, 28, 28, 28,
 28, 28, 28, 28, 28, 24, 21, 17, 20, 20, 20, 20, 20, 20, 20, 20,
 20, 22, 24, 26, 28, 28, 28, 28, 28, 28, 28, 28, 28, 26, 24, 22,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 8, 12, 16, 16, 16, 16,
 16, 16, 16, 16, 16, 12, 8, 4, 8, 8, 8, 8, 8, 8, 8, 8,
 8, 10, 12, 14, 16, 16, 16, 16, 16, 16, 16, 16, 16, 14, 12, 10,
 11, 11, 11, 11, 11, 11, 11, 11, 11, 12, 13, 15, 16, 16, 16, 16,
 16, 16, 16, 16, 16, 15, 13, 12, 0, 0, 0, 0, 0, 0, 0, 0};
static char default_blue[256]
=
{0, 42, 0, 42, 0, 42, 0, 42, 21, 63, 21, 63, 21, 63, 21, 63,
 0, 5, 8, 11, 14, 17, 20, 24, 28, 32, 36, 40, 45, 50, 56, 63,
 63, 63, 63, 63, 63, 47, 31, 16, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 16, 31, 47, 63, 63, 63, 63, 63, 63, 63, 63, 63, 55, 47, 39,
 31, 31, 31, 31, 31, 31, 31, 31, 31, 39, 47, 55, 63, 63, 63, 63,
 63, 63, 63, 63, 63, 58, 54, 49, 45, 45, 45, 45, 45, 45, 45, 45,
 45, 49, 54, 58, 63, 63, 63, 63, 28, 28, 28, 28, 28, 21, 14, 7,
 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 14, 21, 28, 28, 28, 28,
 28, 28, 28, 28, 28, 24, 21, 17, 14, 14, 14, 14, 14, 14, 14, 14,
 14, 17, 21, 24, 28, 28, 28, 28, 28, 28, 28, 28, 28, 26, 24, 22,
 20, 20, 20, 20, 20, 20, 20, 20, 20, 22, 24, 26, 28, 28, 28, 28,
 16, 16, 16, 16, 16, 12, 8, 4, 0, 0, 0, 0, 0, 0, 0, 0,
 0, 4, 8, 12, 16, 16, 16, 16, 16, 16, 16, 16, 16, 14, 12, 10,
 8, 8, 8, 8, 8, 8, 8, 8, 8, 10, 12, 14, 16, 16, 16, 16,
 16, 16, 16, 16, 16, 15, 13, 12, 11, 11, 11, 11, 11, 11, 11, 11,
 11, 12, 13, 15, 16, 16, 16, 16, 0, 0, 0, 0, 0, 0, 0, 0};

void
delay (void)
{
  int i;
  for (i = 0; i < 10; i++);
}

/* Modifications to restorepalette to allow custom palette setting from a file.
 * The format of the file is simply a table of color number -> color mappings:
 * <color(0-255)> <red(0-63)> <green(0-63)> <blue(0-63)> <rest of line ignored>
 *                       By Charles Blake, chuckb@alice.wonderland.caltech.edu
 */

void
process_palette_file (char **argv)
{
  int i, n, r, g, b;
  FILE *palette_file;

  if ((palette_file = fopen (argv[1], "r")) == NULL)
    {
      fprintf (stderr, "%s opening %s: %s\n", argv[0], argv[1], strerror (errno));
      exit (1);
    }
/*
 * NOTE:The tricky %*[^\n] conversion specifier allows arbitrary text following
 * any line with a valid entry on it. This is a simple comment syntax device.
 */
  for (n = 1; fscanf (palette_file, "%d %d %d %d %*[^\n]", &i, &r, &g, &b) == 4; n++)
    {
      default_red[i] = r;	/* deferred assignment prevents partial */
      default_green[i] = g;	/* assignment in case of in-line errors */
      default_blue[i] = b;
    }
  if (!feof (palette_file))
    {
      fprintf (stderr, "%s parse error: %s: line %d\n", argv[0], argv[1], n);
      exit (1);
    }
}

int
main (int argc, char *argv[])
{
  int i;

  vga_init ();
  if (argc > 1)
    process_palette_file (argv);

  if (vga_getcurrentchipset () != EGA)
    {
      /* Restore textmode/16-color mode palette */
		if (vga_getcurrentchipset() != FBDEV)
      for (i = 0; i < 16; i++)
	{
	  port_in (IS1_RC);
	  delay ();
	  port_out (i, ATT_IW);	/* set palette color number */
	  delay ();
	  port_out (i, ATT_IW);	/* set palette color value */
	  delay ();
	}

      /* Restore 256-color VGA RGB look-up table */
      for (i = 0; i < 256; i++)
	vga_setpalette (i, default_red[i], default_green[i],
			default_blue[i]);

      if (vga_getcurrentchipset() != FBDEV) {
      port_in (IS1_RC);
      delay ();
      port_out (0x20, ATT_IW);	/* enable display */
      }
    }
  return 0;
}
